core/new-gui/src/app/workspace/service/workflow-graph/model/workflow-action.service.ts

```diff
@@ -44,6 +45,51 @@ type GroupInfo = {
   layer: number;
 };
 
+// Caveat: These operations must be performed in order, and the ability to sync up multiple clients relies
+// on the fact that different clients rest at the same state.
+
+// At least for some of them, have to do a bit more thinking.
+
+export type commandFuncs =
+  | "undoredo"
+  | "addOperator"
+  | "deleteOperator"
+  | "addOperatorsAndLinks"
+  | "deleteOperatorsAndLinks"
+  | "changeOperatorPosition"
+  | "autoLayoutWorkflow"
+  | "setOperatorProperty"
+  | "addLink"
+  | "deleteLink"
+  | "deleteLinkWithID"
+  | "resetAsNewWorkflow";
+
+// keyof yields permitted property names for T. When we pass function, it'll return value of that function?
+// For this type, we index T with the property names for T, which results in us getting the values.
```

Comment:
as those type/interface declaration becomes more and more. I think we should consider moving them to some other files instead of declaring them inside this service file.

Reply:
These types are bounded with `workflowActionService`, as do some types before this PR. If we want to do such refactoring we can do it later along with the existing types.

Reply:
sure. please keep a note/comment and then resolve this.

Reply:
I added a TODO for this.